### PR TITLE
support wakapi url or custom url, fix Manual date error

### PR DESCRIPTION
**Pull request by Wanxp**
1. add support for wakapi server url
2. add support for custom wakapi or other wakatime-api-implement server url
3. fix bug: if not change anything in ManualModal , it will throw an error
4. add support if User input wrong date, it will throw an error

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -322,17 +344,111 @@ class WakaBoxSettingTab extends PluginSettingTab {
...
+					await this.plugin.saveSettings();
+				}).inputEl.size = 60);
+	}
+
+	private async testConnection() {
+		const yesterday = moment().subtract(1, 'days').format("YYYY-MM-DD");
+		const now = moment().format("YYYY-MM-DD");
+		const url = this.plugin.settings.apiUrl + `/users/current/summaries?start=${yesterday}&end=${now}&api_key=` + this.plugin.settings.apiKey;
+		try {
+			const result = await request(url);
```
Comment (simonla) [inline src/main.ts]: 'result' is assigned a value but never used

File: src/main.ts [hunk 2]
```diff
@@ -366,11 +483,29 @@ export class ManualModal extends Modal {
...
+	private checkInput(result: string):boolean {
+		let dateResult = "";
+		try {
+			dateResult = moment(result).format("YYYY-MM-DD");
+		} catch (e) {
+			new Notice(`WakaTime box: fail due to ${e}`, 5000);
+			return false;
+		}
+		if (dateResult == 'Invalid date') {
+			new Notice(`WakaTime box: fail due to Invalid date`, 5000);
```
Comment (simonla) [inline src/main.ts]: [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

File: src/main.ts [hunk 3]
```diff
@@ -322,17 +344,111 @@ class WakaBoxSettingTab extends PluginSettingTab {
...
+				}).inputEl.size = 60)
+			.setDisabled(true);
+	}
+
+	private  createCustomUrlInputSetting(apiUrl: HTMLDivElement) {
+		apiUrl.empty();
+
+		new Setting(apiUrl)
+			.setName('WakaTime API Url Input')
+			.setDesc('Enter your own Wakapi API Url Or Other Server, Attention: "/api/compat/wakatime/v1" is required if you use your own wakapi server')
```
Comment (simonla) [inline src/main.ts]: [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

File: src/main.ts [hunk 4]
```diff
@@ -322,17 +344,111 @@ class WakaBoxSettingTab extends PluginSettingTab {
...
+						.onChange(async (value) => {
+							this.plugin.settings.apiKey = value;
+							await this.plugin.saveSettings();
+						});
+					text.inputEl.size = 60;
+				}
+
+			);
+		new Setting(apiUrlType)
+			.setName('WakaTime API Url Server Type')
```
Comment (simonla) [inline src/main.ts]: [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

File: src/main.ts [hunk 5]
```diff
@@ -322,17 +344,111 @@ class WakaBoxSettingTab extends PluginSettingTab {
...
+
+			});
+		if (this.plugin.settings.apiUrlType != ApiUrlType.Custom) {
+			this.createWakaUrlSettingShow(apiUrl);
+		} else {
+			this.createCustomUrlInputSetting(apiUrl);
+		}
+
+		new Setting(containerEl)
+			.setName('Test Connection')
```
Comment (simonla) [inline src/main.ts]: [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

File: src/main.ts [hunk 6]
```diff
@@ -322,17 +344,111 @@ class WakaBoxSettingTab extends PluginSettingTab {
...
+				.onClick(async () => {
+					await this.testConnection();
+				}));
+	}
+
+	private createWakaUrlSettingShow(apiUrl: HTMLDivElement) {
+		apiUrl.empty();
+		new Setting(apiUrl)
+			.setName('WakaTime API Url')
+			.setDesc('The WakaTime API Url being used now')
```
Comment (simonla) [inline src/main.ts]: [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

**Review by simonla — CHANGES_REQUESTED** [review]

File: src/main.ts (hunk 2 above)
Comment (Wanxp) [reply src/main.ts]: result has removed

File: src/main.ts (hunk 3 above)
Comment (Wanxp) [reply src/main.ts]: Desc change to `Enter your own Wakapi API url or other server, attention: "/api/compat/wakatime/v1" is required if you use your own Wakapi server'`

File: src/main.ts (hunk 4 above)
Comment (Wanxp) [reply src/main.ts]: Name change to `WakaTime API url server type`

File: src/main.ts (hunk 6 above)
Comment (Wanxp) [reply src/main.ts]: Desc changed to `The WakaTime API url being used now`

File: src/main.ts (hunk 5 above)
Comment (Wanxp) [reply src/main.ts]: Name changed to `Test Connection`

File: src/main.ts (hunk 2 above)
Comment (Wanxp) [reply src/main.ts]: Notice changed to `WakaTime box: fail due to invalid date`

**Comment by simonla** [comment]
The `current server url` displayed in the settings sometimes does not match the `server type`. Especially after relaunch obsidian. So I temporarily rebase the master.

**Pull request merged**